R/g.getstarttime.R

```diff
@@ -1,146 +1,140 @@
-g.getstarttime = function(datafile, P, header, mon, dformat, desiredtz, configtz = NULL) {
-  #get input variables (relevant when read.myacc.csv is used)
-  #------------------------------------------------------------
-  if (mon == MONITOR$AXIVITY && dformat == FORMAT$CWA) {
-    starttime = P$data[1,1]
-    starttime = as.POSIXlt(starttime, tz = desiredtz, origin = "1970-01-01")
-  } else if (mon == MONITOR$GENEACTIV && dformat == FORMAT$BIN) {
-    starttime = as.POSIXlt(P$data.out$time[1], tz = desiredtz, origin = "1970-01-01")
-  } else if (dformat == FORMAT$CSV && (mon == MONITOR$ACTIGRAPH || mon == MONITOR$AXIVITY || mon == MONITOR$VERISENSE)) {
-    if (mon == MONITOR$ACTIGRAPH || mon == MONITOR$VERISENSE) {
-      tmph = read.csv(datafile, nrow = 8, skip = 1)
-      tmphi = 1
-      while (tmphi < 10) {
-        if (length(unlist(strsplit(format(tmph[tmphi,1]),"Start Time"))) > 1) {
-          break
-        }
-        tmphi = tmphi + 1
+g.getstarttime = function(datafile, data, mon, dformat, desiredtz, configtz = NULL) {
+  starttime = NULL
+  if (is.null(configtz)) {
+    configtz = desiredtz
+  }
+  if ("time" %in% colnames(data)) {
+    starttime = as.POSIXlt(data[1, "time"], tz = desiredtz, origin = "1970-01-01")
+  } else if (mon == MONITOR$MOVISENS) {
+      starttime = unisensR::readUnisensStartTime(dirname(datafile))
+
+      # movisens timestamp is stored in unisens.xml file as a string "%Y-%m-%dT%H:%M:%S",
+      # without a timezone.
+      # unisensR::readUnisensStartTime() converts this string into a POSIXct object,
+      # but since no timezone information was stored, this conversion happens using
+      # the system timezone.
+      # If configtz is different from the system timezone (""), we need to force the
+      # correct timezone (force, while keeping the same hh:mm:ss time, not just convert to congigtz)
+      if (configtz != "") {
+        starttime = lubridate::force_tz(starttime, configtz)
```

Comment:
In this particular module, I could indeed go back to forcing the timezone with these two lines:
```
starttime = as.POSIXlt(starttime)
starttime = as.POSIXct(starttime, tz=configtz)
```

This two-liner does accomplish the same as lubridate::force_tz().

- (only using `as.POSIXct(starttime, tz=configtz)` wouldn't accomplish the job because it just changes the timezone attribute, without changing teh underlying Unix timestamp. 

    What we need is to change the underlying Unix timestamp in such a way that the the new timestamp will have the same hh:mm:ss value with the new timezone, as the old timestamp did with the old timezone.

    For example, if we had a timestamp "2020-08-26 10:09:00 GMT" and configtz="America/New_York",
we'd want it to become
"2020-08-26 10:09:00 EDT" (same hh:mm:ss value, different timezone), not "2020-08-26 06:09:00 EDT", which is what we'd get with a normal timestamp conversion using `as.POSIXct(..., tz=configtz)`)

I was trying to avoid bringing in `lubridate`, so I originally used the `as.POSIXlt/as.POSIXct` combo, but it turns out, as.POSIXlt() is *super* expensive. Just crazy expensive, nothing that could be compromised with.

As you correctly noted, in this particular module we are only forcing the timezone on one timestamp, so the speed doesn't matter at all, so I definitely can get rid of `lubridate` here. However, there are two paces in `g.readaccfile` where the same timestamp correction is needed ([this commit](https://github.com/wadpac/GGIR/commit/6504623a93ee344261475dd9deef7b78050c8c1a) touches all the relevant code), and there it's done on an entire list of timestamps, and we really feel the cost of as.POSIXlt() there. It was really bad, part 1 for gt3x files got about 10 times slower.

So I figured if we already need `lubridate` for `g.readaccfile`, why not use it in  `g.getstarttime` as well.

And I don't think there's a good way to avoid using `lubridate` in `g.readaccfile`. If we revert to the way timestamps for gt3x and Axivity csv files were handled prior to this PR, we accept that specifying `configtz` simply doesn't work for these files. For these two file types, we need a way to force a timezone, and as far as I can tell there are only two options, the crazy-slow `as.POSIXlt/as.POSIXct` combo, and  `lubridate::force_tz()`

Please let me know what you think, or if you see another solution to this issue!

Reply:
I follow you, in that case let's keep lubridate and use it in those cases you list.